source/src/ble/cs_CrownstoneCentral.cpp

```diff
@@ -0,0 +1,663 @@
+		return false;
+	}
+	if (_currentStep != step) {
+		LOGCsCentralInfo("Wrong step: _currentStep=%u step=%u", _currentStep, step);
+		finalizeOperation(_currentOperation, ERR_WRONG_OPERATION);
+		return false;
+	}
+	return true;
+}
+
+void CrownstoneCentral::finalizeOperation(Operation operation, cs_ret_code_t retCode) {
+	switch (operation) {
+		case Operation::NONE:
+		case Operation::CONNECT: {
+			finalizeOperation(operation, reinterpret_cast<uint8_t*>(&retCode), sizeof(retCode));
+			break;
+		}
+		case Operation::WRITE: {
+			cs_central_write_result_t result = {
+					.writeRetCode = retCode,
+					.result = ResultPacketAccessor<>()
+			};
+			finalizeOperation(operation, reinterpret_cast<uint8_t*>(&result), sizeof(result));
+			break;
+		}
+	}
+}
+
+void CrownstoneCentral::finalizeOperation(Operation operation, uint8_t* data, uint8_t dataSize) {
+	LOGCsCentralDebug("finalizeOperation operation=%u _currentOperation=%u dataSize=%u", operation, _currentOperation, dataSize);
+	event_t event(CS_TYPE::CONFIG_DO_NOT_USE);
+
+	// Handle error first.
+	if (_currentOperation != operation) {
+		switch (_currentOperation) {
+			case Operation::NONE: {
+				break;
+			}
+			case Operation::CONNECT: {
+				// Ignore data, finalize with error instead.
+				TYPIFY(EVT_CS_CENTRAL_CONNECT_RESULT) result = ERR_WRONG_STATE;
+				event_t errEvent(CS_TYPE::EVT_CS_CENTRAL_CONNECT_RESULT, reinterpret_cast<uint8_t*>(&result), sizeof(result));
+				sendOperationResult(errEvent);
+				return;
+			}
+			case Operation::WRITE: {
+				// Ignore data, finalize with error instead.
+				TYPIFY(EVT_CS_CENTRAL_WRITE_RESULT) result = {
+						.writeRetCode = ERR_WRONG_STATE,
+						.result = ResultPacketAccessor<>()
+				};
+				event_t errEvent(CS_TYPE::EVT_CS_CENTRAL_WRITE_RESULT, reinterpret_cast<uint8_t*>(&result), sizeof(result));
+				sendOperationResult(errEvent);
+				return;
+			}
+		}
+	}
+
+	event.data = data;
+	event.size = dataSize;
+	switch (_currentOperation) {
+		case Operation::NONE: {
+			LOGCsCentralDebug("No operation was in progress");
+			break;
+		}
+		case Operation::CONNECT: {
+			event.type = CS_TYPE::EVT_CS_CENTRAL_CONNECT_RESULT;
+			break;
+		}
+		case Operation::WRITE: {
+			event.type = CS_TYPE::EVT_CS_CENTRAL_WRITE_RESULT;
+			break;
+		}
+	}
+	sendOperationResult(event);
+}
+
+void CrownstoneCentral::sendOperationResult(event_t& event) {
+	LOGCsCentralDebug("sendOperationResult type=%u size=%u", event.type, event.size);
+	// Set current operation before dispatching the event, so that a new command can be issued on event.
+	_currentOperation = Operation::NONE;
+	_currentStep = 0;
+
+	if (event.type != CS_TYPE::CONFIG_DO_NOT_USE) {
+		event.dispatch();
+	}
+}
+
+//////////////////// Handlers ////////////////////
+
+void CrownstoneCentral::onConnect(cs_ret_code_t retCode) {
+	if (!finalizeStep(ConnectSteps::CONNECT, retCode)) {
+		return;
+	}
+
+	retCode = BleCentral::getInstance().discoverServices(_uuids, sizeof(_uuids) / sizeof(_uuids[0]));
```

Comment:
We have a macro `SIZEOF_ARRAY` in `cs_Utils.h`.
I suggest reimplementing that as:
```
template<class T, size_t N>
constexpr auto ArraySize(T(&)[N]) { return N; } 
```

Reply:
Oh I didn't even think about using a function for that :P Seems like more work to include the utils for that than to write the calculation.

Your suggestion is very unreadable to me. How does it work, and why is it better than the macro?

Reply:
Usage of the suggested template is safer because the compiler can detect if you are using it on an invalid type (such as a pointer to void, pointer to vraiable length array etc.)

Usage;
```
uint8_t myArray[4];
ArraySize(myArray); // returns 4;

uint8_t myInvalidArray[] = new Array[arrayLen];
ArraySize(myInvalidArray); // compilation error.
```